Packages/com.unity.multiplayer.samples.coop/Utilities/EditorChildSceneLoader.cs

```diff
@@ -0,0 +1,101 @@
+using System;
+using System.Collections.Generic;
+#if UNITY_EDITOR
+using UnityEditor;
+using UnityEditor.SceneManagement;
+#endif
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+/// <summary>
+/// Allows setting a scene as a root scene and setting its child scenes. To use this, drag this component on any object in a scene to make that scene a root scene. In the background, ChildSceneLoader will automatically manage this.
+/// </summary>
+public class EditorChildSceneLoader : MonoBehaviour
+{
+#if UNITY_EDITOR
+    [SerializeField]
+    public List<SceneAsset> ChildScenesToLoadConfig;
+
+    void Update()
+    {
+        // DO NOT DELETE keep this so we can enable/disable this script... (used in ChildSceneLoader)
+    }
+
+    public void SaveSceneSetup()
+    {
+        ChildScenesToLoadConfig ??= new();
+        ChildScenesToLoadConfig.Clear();
+        foreach (var sceneSetup in EditorSceneManager.GetSceneManagerSetup())
+        {
+            ChildScenesToLoadConfig.Add(AssetDatabase.LoadAssetAtPath<SceneAsset>(sceneSetup.path));
+        }
+    }
+
+    public void ResetSceneSetupToConfig()
+    {
+        var sceneAssetsToLoad = ChildScenesToLoadConfig;
+
+        List<SceneSetup> sceneSetupToLoad = new();
+        foreach (var sceneAsset in sceneAssetsToLoad)
+        {
+            sceneSetupToLoad.Add(new SceneSetup() { path = AssetDatabase.GetAssetPath(sceneAsset), isActive = false, isLoaded = true });
+        }
+
+        sceneSetupToLoad[0].isActive = true;
+        EditorSceneManager.SaveCurrentModifiedScenesIfUserWantsTo();
+        EditorSceneManager.RestoreSceneManagerSetup(sceneSetupToLoad.ToArray());
+    }
+#endif
+}
+
+#if UNITY_EDITOR
+[CustomEditor(typeof(EditorChildSceneLoader))]
+public class ChildSceneLoaderInspectorGUI : Editor
+{
+    public override void OnInspectorGUI()
+    {
+        base.OnInspectorGUI();
+
+        var currentInspectorObject = (EditorChildSceneLoader)target;
+
+        if (GUILayout.Button("Save scene setup to config"))
+        {
+            currentInspectorObject.SaveSceneSetup();
+        }
+
+        if (GUILayout.Button("Reset scene setup from config..."))
+        {
+            currentInspectorObject.ResetSceneSetupToConfig();
+        }
+    }
+}
+
+[InitializeOnLoad]
+public class ChildSceneLoader
+{
+    static ChildSceneLoader()
+    {
+        EditorSceneManager.sceneOpened += OnSceneLoaded;
+    }
+
+    static void OnSceneLoaded(Scene _, OpenSceneMode mode)
+    {
+        if (mode != OpenSceneMode.Single || BuildPipeline.isBuildingPlayer) return; // try to load child scenes only for root scenes or if not building
+
+        var scenesToLoadObjects = GameObject.FindObjectsOfType<EditorChildSceneLoader>();
```

Comment:
I'm not happy with this, I need to analyze the perf for this, this will be an issue, especially for more complex scenes

Reply:
Just profiled, time spent here is super small compared to loading times. This is fine...